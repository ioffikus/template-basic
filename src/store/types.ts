import { RouterState } from 'connected-react-router';
import { IAuthState } from 'src/ducks/auth';
import { ICartState } from 'src/ducks/cart';
import { ICardPaidItemsState } from 'src/ducks/cartPaidItems';
import { ICatalogQueryState } from 'src/ducks/catalogQuery';
import { ICatalogState } from 'src/ducks/catalog';
import { ICategoriesState } from 'src/ducks/categories';
import { ICategoriesMenuState } from 'src/ducks/categoriesMenu';
import { IFilterState } from 'src/ducks/filter';
import { ILocaleState } from 'src/ducks/locale';
import { IRoutesState } from 'src/ducks/routes';
import { IMetaState } from 'src/ducks/meta';
import { IPageState } from 'src/ducks/page';
import { IProductState } from 'src/ducks/product';
import { IProgressState } from 'src/ducks/progress';
import { IRecommendItemsState } from 'src/ducks/recommendItems';
import { IPopularItemsState } from 'src/ducks/popularItems';
import { IUIState } from 'src/ducks/ui';
import { IUserState } from 'src/ducks/user';
import { IUserDeliveryState } from 'src/ducks/userDelivery';
import { IUserOrdersState } from 'src/ducks/userOrders';
import { Dispatch } from 'redux';
import { IXShopHeaders } from 'src/core/interfaces/IXShopHeaders';

export interface IState {
  router?: RouterState;
  ui?: IUIState;
  auth?: IAuthState;
  user?: IUserState;
  categories?: ICategoriesState;
  categoriesMenu?: ICategoriesMenuState;
  catalogQuery?: ICatalogQueryState;
  catalog?: ICatalogState;
  product?: IProductState;
  progress?: IProgressState;
  routes?: IRoutesState;
  meta?: IMetaState;
  filter?: IFilterState;
  cart?: ICartState;
  cartPaidItems?: ICardPaidItemsState;
  userDelivery?: IUserDeliveryState;
  userOrders?: IUserOrdersState;
  recommendItems?: IRecommendItemsState;
  popularItems?: IPopularItemsState;
  page?: IPageState;
  locale?: ILocaleState;
}

export interface IReduxCommonState {
  isRequestError: boolean;
  isRequesting: boolean;
}

export interface IFetchDataFn {
  (abortToken?: string): Promise<any>;
}

export interface IReduxThunkCb {
  (dispatch: Dispatch<Store.IState>, getState: () => Store.IState, xShopHeaders?: IXShopHeaders): Promise<any>;
}
