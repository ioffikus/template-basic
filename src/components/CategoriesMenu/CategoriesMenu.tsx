import React from 'react';
import { connect } from 'react-redux';
import { Menu } from 'antd';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import PropTypes from 'prop-types';
import queryString from 'query-string';
import getCategoriesPath from 'src/helpers/getCategoriesPath';

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  categories: Alicanto.Models.Category[];
  selectedId: number;
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  openKeys: string[];
  selectedKeys: string[];
}

const mapStateToProps = (state: Store.IState) => ({
  ui: state.ui,
  categories: state.categoriesMenu.content,
  selectedId: _.get(state.catalogQuery.filterObj, 'categories__in[0]', null),
});

const WrappedCategoriesMenu = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class CategoriesMenu extends React.Component<IProps, IState> {
    static childContextTypes = {
      siderCollapsed: PropTypes.bool,
    };

    getChildContext() {
      return {
        siderCollapsed: false,
      };
    }

    setKeys = (selectedId: number) => {
      const keys: { openKeys: string[]; selectedKeys: string[] } = {
        openKeys: [],
        selectedKeys: [],
      };
      if (!_.isNull(selectedId)) {
        const categoriesPath = getCategoriesPath(this.props.categories, selectedId);
        if (!_.isEmpty(categoriesPath)) {
          keys.openKeys = categoriesPath.map(category => '' + category.id);
          keys.selectedKeys = ['' + keys.openKeys.pop()];
        }
      }
      this.setState({ ...keys });
    };

    componentWillMount() {
      this.setKeys(this.props.selectedId);
    }

    componentWillReceiveProps(nextProps: IProps) {
      if (!_.isEqual(this.props.selectedId, nextProps.selectedId)) {
        this.setKeys(nextProps.selectedId);
      }
    }

    onOpenChange = (openKeys: string[]) => {
      this.setState({ openKeys });
    };

    onSelect = ({ selectedKeys }: { selectedKeys: string[] }) => {
      this.setState({ selectedKeys });
    };

    onDeselect = ({ selectedKeys }: { selectedKeys: string[] }) => {
      this.setState({ selectedKeys });
    };

    render() {
      const categoriesMenuUI = this.props.ui.common.categoriesMenu;
      const renderMenuItems = (categories: Alicanto.Models.Category[]): React.ReactNode[] => {
        return categories.map(category => {
          if (!_.isEmpty(category.subcategories)) {
            return (
              <Menu.SubMenu key={category.id} title={category.name}>
                {renderMenuItems(category.subcategories)}
              </Menu.SubMenu>
            );
          }
          return (
            <Menu.Item key={category.id}>
              <Link
                to={{
                  pathname: '/catalog',
                  search: queryString.stringify({
                    filter: JSON.stringify({
                      categories__in: [category.id],
                    }),
                  }),
                }}
              >
                {category.name}
              </Link>
            </Menu.Item>
          );
        });
      };
      return (
        <div className="root app-main-menu">
          <style jsx>{`
            .root {
              padding-left: 16px;
            }
            .root :global(.footer) {
              position: fixed;
              bottom: 20px;
              left: 0;
              width: 260px;
              color: #b2bdcb;
              font-size: 6px;
              text-align: center;
              white-space: nowrap;
              overflow: hidden;
              transition: 0.2s;
            }
          `}</style>
          <Menu
            mode="inline"
            openKeys={this.state.openKeys}
            selectedKeys={this.state.selectedKeys}
            onOpenChange={this.onOpenChange}
            onSelect={this.onSelect}
            onDeselect={this.onDeselect}
            inlineIndent={16}
            inlineCollapsed={false}
          >
            {renderMenuItems(this.props.categories)}
          </Menu>
          <div className="footer">{categoriesMenuUI.copyright.text.i18n.defaultText}</div>
        </div>
      );
    }
  },
);

export default WrappedCategoriesMenu;
