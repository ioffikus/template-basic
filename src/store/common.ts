import { AxiosError } from 'axios';
import _ from 'lodash';
import { message } from 'antd';
import { BaseError } from 'make-error';

export class BusinessLogicError extends BaseError {}

const beautifyMessage = (message: object) => {
  try {
    return JSON.stringify(message).replace(/"/g, '');
  } catch (e) {
    return 'unknown message text';
  }
};

export const showError = (error: AxiosError | BusinessLogicError) => {
  if (process.env.BROWSER) {
    if (error instanceof BusinessLogicError) {
      message.error(error.message);
      return;
    }
    const responseStatus = _.get(error, 'response.status', 'unknown status');
    const responseData = _.get(error, 'response.data', null);
    const responseBody = _.get(error, 'response.body', null);
    const simpleError = _.get(responseData, 'error', null);
    const errorMessage = _.get(simpleError, 'message', null);
    const errorDetails = _.get(simpleError, 'detail', null);
    const errorsObject = _.get(simpleError, 'errors', null);
    if (!responseData && !responseBody) {
      message.error(`Status code: ${responseStatus}. Message: unknown message text`);
    } else if (errorMessage) {
      message.error(`Status code: ${responseStatus}. Message: ${errorMessage}`);
    } else if (errorDetails) {
      message.error(`Status code: ${responseStatus}. Message: ${errorDetails}`);
    } else if (errorsObject) {
      message.error(`Status code: ${responseStatus}. Message: ${beautifyMessage(errorsObject)}`);
    } else if (simpleError) {
      message.error(`Status code: ${responseStatus}. Message: ${beautifyMessage(simpleError)}`);
    } else if (!_.isNull(responseData)) {
      message.error(`Status code: ${responseStatus}. Message: ${beautifyMessage(responseData)}`);
    } else if (!_.isNull(responseBody)) {
      message.error(`Status code: ${responseStatus}. Message: ${beautifyMessage(responseBody)}`);
    } else {
      message.error(`Status code: ${responseStatus}. Message: ${beautifyMessage(error)}`);
    }
  }
};

// WIP deprecated
// export const parseAxiosResponseToError = (resp: AxiosResponse): string => {
//   const responseStatus = _.get(resp, 'status', 'unknown status');
//   let messageText;
//   try {
//     messageText = JSON.stringify(_.get(resp, 'data', {}));
//   } catch (e) {
//     messageText = e;
//   }
//   return `Status code: ${responseStatus}. Message: ${messageText}`;
// };
