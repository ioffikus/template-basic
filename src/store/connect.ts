import { Dispatch, connect } from 'react-redux';

interface MapPropsParam<TProps> {
  (state: Store.IState, ownProps?: TProps): TProps;
}

interface MapDispatchParam<TProps, TDispatchProps> {
  (dispatch: Dispatch<Store.IState>, ownProps?: TProps): TDispatchProps;
}

export interface ConnectedComponent<TProps> {
  <TComponent extends React.ComponentType<TProps>>(component: TComponent): TComponent;
}

export default function<TProps, TDispatchProps = {}>(
  mapProps: MapPropsParam<TProps>,
  mapDispatch?: MapDispatchParam<TProps, TDispatchProps>,
): ConnectedComponent<TProps> {
  return connect<TProps, TDispatchProps, TProps>(mapProps, mapDispatch) as ConnectedComponent<TProps & TDispatchProps>;
}

// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/9951#issuecomment-317226870
