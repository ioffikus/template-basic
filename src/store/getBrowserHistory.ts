import H, { createBrowserHistory } from 'history';
import _ from 'lodash';

let history: H.History = null;

const getBrowserHistory = () => {
  if (_.isNull(history)) {
    history = createBrowserHistory();
  }
  return history;
};

export default getBrowserHistory;
