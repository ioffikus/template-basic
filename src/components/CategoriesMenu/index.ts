import CategoriesMenu from './CategoriesMenu';

export default CategoriesMenu;
