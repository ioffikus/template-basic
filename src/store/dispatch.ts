import { Store } from 'redux';
import _ from 'lodash';
import { showError } from 'src/store/common';

let store: Store<Store.IState>;

export const setupDispatch = (config: { store: Store<Store.IState> }) => {
  store = config.store;
};

const dispatch = (action: any) => {
  if (_.isFunction(action)) {
    return store.dispatch(action).catch((error: any) => {
      if (!_.isNull(error)) {
        showError(error);
      }
      return Promise.reject(error);
    });
  }
  return store.dispatch(action);
};

export default dispatch;
