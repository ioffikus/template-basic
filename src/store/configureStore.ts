import { connectRouter, routerMiddleware } from 'connected-react-router';
import { StaticRouter } from 'react-router-dom';
import { StaticRouterProps } from 'react-router';
import { applyMiddleware, compose, createStore, Store } from 'redux';
import thunk from 'redux-thunk';
import { types } from 'redux-act';
import H from 'history';
import _ from 'lodash';
import DevTools from 'src/components/DevTools';
import ConfigHelper from 'src/helpers/configHelper';
import rootReducer from 'src/ducks';
import { IXShopHeaders } from 'src/core/interfaces/IXShopHeaders';
import getBrowserHistory from 'src/store/getBrowserHistory';
import { setupDispatch } from 'src/store/dispatch';

interface IConfigureStore {
  (initialState: Store.IState, SSRParams?: { location: string; xShopHeaders: IXShopHeaders }): Store<Store.IState>;
}

const configureStore: IConfigureStore = (initialState = {}, SSRParams) => {
  let history: H.History;
  let middlewares;

  if (process.env.BROWSER) {
    history = getBrowserHistory();
    middlewares = [routerMiddleware(history)];
    middlewares.push(thunk);
  } else {
    const staticRouter: React.Component<StaticRouterProps> = new StaticRouter();
    const location = _.get(SSRParams, 'location', '');
    staticRouter.props = { location, context: {}, basename: '' };
    history = _.get(staticRouter.render(), 'props.history');
    middlewares = [routerMiddleware(history)];
    const xShopHeaders = _.get(SSRParams, 'xShopHeaders');
    middlewares.push(thunk.withExtraArgument(xShopHeaders));
  }

  const store = createStore(
    connectRouter(history)(rootReducer),
    initialState,
    ConfigHelper.isDev() && process.env.BROWSER
      ? compose(applyMiddleware(...middlewares), DevTools.instrument())
      : compose(applyMiddleware(...middlewares)),
  );

  if (ConfigHelper.isDev() && module.hot) {
    module.hot.accept('src/ducks', () => {
      types.clear();
      const nextRootReducer = require('src/ducks/index').default;
      store.replaceReducer(connectRouter(history)(nextRootReducer));
    });
  }

  setupDispatch({ store });

  return store;
};

export default configureStore;
