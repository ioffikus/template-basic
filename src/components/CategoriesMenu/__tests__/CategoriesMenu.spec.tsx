import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import { setupDispatch } from 'src/store/dispatch';
import uiStateMock from 'src/store/__mocks__/uiStore';
import categoriesMenuStateMock from 'src/store/__mocks__/categoriesMenuStore';
import catalogQueryStateMock from 'src/store/__mocks__/catalogQueryStore';
import CategoriesMenu from '../CategoriesMenu';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe(null, () => {
  it('render() correct', () => {
    const store = mockStore({
      ui: uiStateMock,
      categoriesMenu: categoriesMenuStateMock,
      catalogQuery: catalogQueryStateMock,
    });
    setupDispatch({ store });
    const wrapper = shallow(<CategoriesMenu />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
