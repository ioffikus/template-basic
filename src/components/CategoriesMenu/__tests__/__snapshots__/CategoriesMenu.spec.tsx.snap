// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() correct 1`] = `
<div
  className="jsx-801695999 root app-main-menu"
>
  <JSXStyle
    css=".root.jsx-801695999{padding-left:16px;}.root.jsx-801695999 .footer{position:fixed;bottom:20px;left:0;width:260px;color:#b2bdcb;font-size:6px;text-align:center;white-space:nowrap;overflow:hidden;-webkit-transition:0.2s;transition:0.2s;}"
    styleId="801695999"
  />
  <Menu
    className=""
    inlineCollapsed={false}
    inlineIndent={16}
    mode="inline"
    onDeselect={[Function]}
    onOpenChange={[Function]}
    onSelect={[Function]}
    openKeys={
      Array [
        "35",
      ]
    }
    prefixCls="ant-menu"
    selectedKeys={
      Array [
        "37",
      ]
    }
    theme="light"
  >
    <MenuItem
      key="23"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B23%5D%7D",
          }
        }
      >
        Adventure novel
      </Link>
    </MenuItem>
    <SubMenu
      key="35"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTitleClick={[Function]}
      onTitleMouseEnter={[Function]}
      onTitleMouseLeave={[Function]}
      title="books"
    >
      <MenuItem
        key="37"
      >
        <Link
          replace={false}
          to={
            Object {
              "pathname": "/catalog",
              "search": "filter=%7B%22categories__in%22%3A%5B37%5D%7D",
            }
          }
        >
          eBooks
        </Link>
      </MenuItem>
      <MenuItem
        key="36"
      >
        <Link
          replace={false}
          to={
            Object {
              "pathname": "/catalog",
              "search": "filter=%7B%22categories__in%22%3A%5B36%5D%7D",
            }
          }
        >
          Paperback
        </Link>
      </MenuItem>
    </SubMenu>
    <MenuItem
      key="24"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B24%5D%7D",
          }
        }
      >
        Brit lit
      </Link>
    </MenuItem>
    <MenuItem
      key="26"
    >
      <Link
        replace={false}
        to={
          Object {
            "pathname": "/catalog",
            "search": "filter=%7B%22categories__in%22%3A%5B26%5D%7D",
          }
        }
      >
        Literary fiction
      </Link>
    </MenuItem>
  </Menu>
  <div
    className="jsx-801695999 footer"
  >
    Alicanto E-commerce Systems
  </div>
</div>
`;
